libsolidity/ast/Types.h

```diff
@@ -1076,6 +1077,56 @@ class EnumType: public Type
 	EnumDefinition const& m_enum;
 };
 
+/**
+ * The type of a UserDefinedValueType.
+ */
+class UserDefinedValueTypeType: public Type
+{
+public:
+	explicit UserDefinedValueTypeType(
+		Type const& _underlyingType,
+		UserDefinedValueTypeDefinition const& _userDefinedValueType
+	):
+		m_underlyingType(_underlyingType),
+		m_userDefinedValueType(_userDefinedValueType)
+	{
+	}
+
+	Category category() const override { return Category::UserDefinedValueType; }
+	Type const& underlyingType() const { return m_underlyingType; }
+	UserDefinedValueTypeDefinition const& userDefinedValueType() const { return m_userDefinedValueType; }
+
+	TypeResult binaryOperatorResult(Token, Type const*) const override { return nullptr; }
+	Type const* encodingType() const override { return &m_underlyingType; }
+	TypeResult interfaceType(bool /* _inLibrary */) const override {return &m_underlyingType; }
+	std::string richIdentifier() const override;
+	bool operator==(Type const& _other) const override;
+
+	unsigned calldataEncodedSize(bool _padded) const override { return m_underlyingType.calldataEncodedSize(_padded); }
+
+	bool leftAligned() const override { return m_underlyingType.leftAligned(); }
+	bool canBeStored() const override { return m_underlyingType.canBeStored(); }
+	u256 storageSize() const override { return m_underlyingType.storageSize(); }
+	bool isValueType() const override
+	{
+		solAssert(m_underlyingType.isValueType(), "");
+		return true;
+	}
+	bool nameable() const override { return m_underlyingType.nameable(); }
+
+	std::string toString(bool _short) const override { return "UserDefinedValueType(" + m_underlyingType.toString(_short) + ")"; }
+
+	BoolResult isExplicitlyConvertibleTo(Type const& _convertTo) const override;
+	///	A helper function to reason about UserDefinedValueType to ValueType explicit conversion
+	static bool convertTo(Type const& _userDefined, Type const& _valueType);
+protected:
+	std::vector<std::tuple<std::string, Type const*>> makeStackItems() const override;
+
```

Comment:
I think we should also overwrite the following two functions:
 - canonicalName (assert false)
 - signatureInExternalFunction (assert false)